File: web/src/lib/job-sources/jsearch.ts
import type { JobPosting } from "../types";

export async function searchJSearch(
  query = "Data Engineer",
  location = "United States",
  postedWithinDays = 7
): Promise<JobPosting[]> {
  const apiKey = process.env.JSEARCH_API_KEY;
  if (!apiKey) return [];

  const datePosted = postedWithinDays <= 7 ? "week" : "month";
  const params = new URLSearchParams({
    query: `${query} in ${location}`,
    page: "1",
    num_pages: "3",
    date_posted: datePosted,
    remote_jobs_only: "false",
  });

  try {
    const resp = await fetch(`https://jsearch.p.rapidapi.com/search?${params}`, {
      headers: {
        "X-RapidAPI-Key": apiKey,
        "X-RapidAPI-Host": "jsearch.p.rapidapi.com",
      },
    });
    if (!resp.ok) return [];
    const data = await resp.json();
    return (data.data || []).map(normalize);
  } catch {
    return [];
  }
}

function normalize(item: Record<string, unknown>): JobPosting {
  const minSal = item.job_min_salary as number | undefined;
  const maxSal = item.job_max_salary as number | undefined;
  let salary = "";
  if (minSal && maxSal) salary = `$${minSal.toLocaleString()} - $${maxSal.toLocaleString()}`;
  else if (minSal) salary = `$${minSal.toLocaleString()}+`;

  return {
    id: `jsearch_${item.job_id || ""}`,
    title: (item.job_title as string) || "",
    company: (item.employer_name as string) || "",
    location: `${item.job_city || ""}, ${item.job_state || ""}`.replace(/^, |, $/g, ""),
    remote: Boolean(item.job_is_remote),
    posted_at: (item.job_posted_at_datetime_utc as string) || null,
    source: "jsearch",
    url: (item.job_apply_link as string) || (item.job_google_link as string) || "",
    jd_text: (item.job_description as string) || "",
    salary,
    visa_signal: "unknown",
    company_domain: (item.employer_website as string) || "",
    raw_data: item,
  };
}

File: job_copilot/app/integrations/jsearch.py
"""JSearch (RapidAPI) job search integration."""
from __future__ import annotations

from datetime import datetime, timedelta
from typing import Any

import httpx

from app.core.config import get_settings
from app.core.logging import get_logger
from app.models.schemas import JobPosting

log = get_logger(__name__)

BASE_URL = "https://jsearch.p.rapidapi.com/search"


async def search_jobs(
    query: str = "Data Engineer",
    location: str = "United States",
    posted_within_days: int = 7,
    page: int = 1,
    num_pages: int = 3,
) -> list[JobPosting]:
    settings = get_settings()
    if not settings.jsearch_api_key:
        log.warning("jsearch_skipped", reason="JSEARCH_API_KEY not set")
        return []

    headers = {
        "X-RapidAPI-Key": settings.jsearch_api_key,
        "X-RapidAPI-Host": "jsearch.p.rapidapi.com",
    }
    date_posted = "week" if posted_within_days <= 7 else "month"
    params = {
        "query": f"{query} in {location}",
        "page": str(page),
        "num_pages": str(num_pages),
        "date_posted": date_posted,
        "remote_jobs_only": "false",
    }

    jobs: list[JobPosting] = []
    async with httpx.AsyncClient(timeout=30) as client:
        try:
            resp = await client.get(BASE_URL, headers=headers, params=params)
            resp.raise_for_status()
            data = resp.json()
            for item in data.get("data", []):
                jobs.append(_normalize(item))
            log.info("jsearch_results", count=len(jobs))
        except Exception as e:
            log.error("jsearch_error", error=str(e))
    return jobs


def _normalize(item: dict[str, Any]) -> JobPosting:
    posted_at = None
    if item.get("job_posted_at_datetime_utc"):
        try:
            posted_at = datetime.fromisoformat(
                item["job_posted_at_datetime_utc"].replace("Z", "+00:00")
            )
        except (ValueError, TypeError):
            pass

    return JobPosting(
        id=f"jsearch_{item.get('job_id', '')}",
        title=item.get("job_title", ""),
        company=item.get("employer_name", ""),
        location=item.get("job_city", "") + ", " + item.get("job_state", ""),
        remote=item.get("job_is_remote", False),
        posted_at=posted_at,
        source="jsearch",
        url=item.get("job_apply_link", "") or item.get("job_google_link", ""),
        jd_text=item.get("job_description", ""),
        salary=_extract_salary(item),
        company_domain=item.get("employer_website", ""),
        raw_data=item,
    )


def _extract_salary(item: dict) -> str:
    min_sal = item.get("job_min_salary")
    max_sal = item.get("job_max_salary")
    if min_sal and max_sal:
        return f"${min_sal:,.0f} - ${max_sal:,.0f}"
    if min_sal:
        return f"${min_sal:,.0f}+"
    return ""
